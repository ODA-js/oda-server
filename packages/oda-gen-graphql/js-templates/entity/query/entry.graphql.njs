  <#@ context 'entity' -#>

  #{entity.pluralEntry}( after: String, first: Int, before: String, last: Int, limit: Int, skip: Int, orderBy: [#{entity.name}SortOrder], filter: #{entity.name}ComplexFilter): #{entity.plural}Connection

  #{entity.singularEntry}(#{entity.unique}): #{entity.name}