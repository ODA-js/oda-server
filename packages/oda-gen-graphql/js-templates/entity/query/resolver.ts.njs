<#@ context 'entity' -#>
import * as log4js from 'log4js';
let logger = log4js.getLogger('graphql:query');
import {get} from 'lodash';

import RegisterConnectors from '../../../../data/registerConnectors';
import { emptyConnection, pagination, detectCursorDirection, consts } from 'oda-api-graphql';
import { lib } from 'oda-gen-common';

const { selectionTree: traverse } = lib;

import { utils } from 'oda-api-graphql';

const { validId } = utils;

export async function fixCount(length: number, cursor: { skip?: number, limit?: number; }, getCount: () => Promise<Number>) {
  const count = await getCount();
  if (count > 0) {
    if (length == cursor.limit) {
      return count
    }
    if (length < cursor.limit) {
      return cursor.skip + length;
    } else {
      return count;
    }
  }
  return count;
}

export const query: { [key: string]: any } = {
  #{entity.plural}: async (
    owner,
    args: {
      after: string,
      first: number,
      before: string,
      last: number,
      orderBy: string | string[],
      filter: object,
      limit: number,
      skip: number,
    },
    context: { connectors: RegisterConnectors },
    info
  ) => {
    logger.trace('#{entity.plural}');
    let result;
    let selectionSet = traverse(info);

    let idMap = {
      id: '#{entity.adapter == 'mongoose' ? '_id' : 'id'}',
<# entity.idMap.forEach(f=>{-#>
      #{f}: '#{f}',
<#})-#>
    };

    let list = get(selectionSet, 'edges.node') ? await context.connectors.#{entity.name}.getList({
      ...args,
      idMap,
    }) : [];

    if (list.length > 0) {
      let cursor = pagination(args);
      let direction = detectCursorDirection(args)._id;

      let edges = get(selectionSet, 'edges') ?
        list.map(l => {
          return {
            cursor: l.id,
            node: l,
          };
        }) : null;

      let pageInfo = get(selectionSet, 'pageInfo') ?
        {
          startCursor: get(selectionSet, 'pageInfo.startCursor')
            ? edges[0].cursor : undefined,
          endCursor: get(selectionSet, 'pageInfo.endCursor')
            ? edges[edges.length - 1].cursor : undefined,
          hasPreviousPage: get(selectionSet, 'pageInfo.hasPreviousPage') ? (direction === consts.DIRECTION.BACKWARD ? list.length === cursor.limit : false) : undefined,
          hasNextPage: get(selectionSet, 'pageInfo.hasNextPage') ? (direction === consts.DIRECTION.FORWARD ? list.length === cursor.limit : false) : undefined,
          count: get(selectionSet, 'pageInfo.count') ?  await fixCount(list.length, cursor, () => context.connectors.#{entity.name}.getCount({
              ...args,
              idMap,
            })) : 0,
        } : null;

      result = {
        edges,
        pageInfo,
      };
    } else {
      result = emptyConnection();
    }
    return result;
  },
  #{entity.singular}: async (
    owner,
    args: {
    <#- for (let f of entity.unique.args) {#>
      #{f.name}?: #{f.type},
    <#-}#>
    <#- for (let f of entity.unique.complex) {
        let args = `${f.fields.map(f=>`${f.name}?: ${f.type}`).join(', ')}`;
      #>
      // #{f.name}
      #{args},
    <#-}#>
    },
    context: { connectors: RegisterConnectors },
    info
  ) => {
    logger.trace('#{entity.singular}');
    let result;
    if (args.id) {
      result = await context.connectors.#{entity.name}.findOneById(args.id);
    <#- for (let f of entity.unique.find) {#>
    } else if (args.#{f.name}) {
      result = await context.connectors.#{entity.name}.findOneBy#{f.cName}(args.#{f.name});
    <#-}#>
    <#- for (let f of entity.unique.complex) {
      let findBy = f.fields.map(f=>f.uName).join('And');
      let loadArgs = `${f.fields.map(f=>`args.${f.name}`).join(', ')}`;
      let condArgs = `${f.fields.map(f=>`args.${f.name}`).join(' && ')}`;
#>
    } else if (#{condArgs}) {
      result = await context.connectors.#{entity.name}.findOneBy#{findBy}(#{loadArgs});
    <#-}#>
    }
    return result;
  },
};