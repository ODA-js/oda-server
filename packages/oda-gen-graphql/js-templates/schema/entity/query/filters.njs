<#@ chunks "$$$main$$$" -#>
<#@ alias 'query-filters'#>
<#@ context 'entity'#>

<#-chunkStart(`./query/filters/${entity.name}Filter.ts`); -#>
import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
    input #{entity.name}Filter {
      or: [#{entity.name}FilterItem]
      and: [#{entity.name}FilterItem]
    <#-entity.filter.forEach((item, index)=>{#>
      #{item}
    <#-})#>
    }
  `,
});

<#-chunkStart(`./query/filters/${entity.name}FilterItem.ts`); -#>
import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
    input #{entity.name}FilterItem {
    <#-entity.filter.forEach((item, index)=>{#>
      #{item}
    <#-})#>
    }
  `,
});

<#-chunkStart(`./query/filters/Embed${entity.name}Filter.ts`); -#>
import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
    input Embed#{entity.name}Filter {
      or: [Embed#{entity.name}FilterItem]
      and: [Embed#{entity.name}FilterItem]
      some: #{entity.name}Filter
      none: #{entity.name}Filter
      every: #{entity.name}Filter
    }
  `,
});

<#-chunkStart(`./query/filters/Embed${entity.name}FilterItem.ts`); -#>
import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
    input Embed#{entity.name}FilterItem {
      some: #{entity.name}Filter
      none: #{entity.name}Filter
      every: #{entity.name}Filter
    }
  `,
});

<#-chunkStart(`./query/filters/${entity.name}FilterSubscriptionsItem.ts`); -#>
import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
    input #{entity.name}FilterSubscriptionsItem {
    <#-entity.filter.forEach((item, index)=>{#>
      #{item}
    <#-})#>
    }
  `,
});


<#-chunkStart(`./query/filters/${entity.name}FilterSubscriptions.ts`); -#>
import { Input } from '../../../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
    input #{entity.name}FilterSubscriptions {
      or: [#{entity.name}FilterSubscriptions]
      and: [#{entity.name}FilterSubscriptions]
      mutation: WhereMutationKind
      node: #{entity.name}FilterSubscriptionsItem
      previous: #{entity.name}FilterSubscriptionsItem
      updatedFields: WhereListOfStrings
    }
  `,
});

<#-chunkStart(`./query/filters/index.ts`); -#>

import Embed#{entity.name}Filter from './Embed#{entity.name}Filter';
import Embed#{entity.name}FilterItem from './Embed#{entity.name}FilterItem';
import #{entity.name}Filter from './#{entity.name}Filter';
import #{entity.name}FilterItem from './#{entity.name}FilterItem';
import #{entity.name}FilterSubscriptions from './#{entity.name}FilterSubscriptions';
import #{entity.name}FilterSubscriptionsItem from './#{entity.name}FilterSubscriptionsItem';
import { Schema } from '../../../../common';

export default new Schema({
  name: '#{entity.name}.queries.filter',
  items: [
    #{entity.name}FilterItem,
    #{entity.name}Filter,
    #{entity.name}FilterSubscriptionsItem,
    #{entity.name}FilterSubscriptions,
    Embed#{entity.name}Filter,
    Embed#{entity.name}FilterItem,
  ],
});
