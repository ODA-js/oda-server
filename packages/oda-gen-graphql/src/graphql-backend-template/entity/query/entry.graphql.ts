import { Entity, ModelPackage, FieldType } from 'oda-model';
import * as inflect from 'inflected';
import { Factory } from 'fte.js';

export const template = 'entity/query/entry.graphql.njs';

export function generate(
  te: Factory,
  entity: Entity,
  pack: ModelPackage,
  role: string,
  aclAllow,
  typeMapper: { [key: string]: (i: FieldType) => string },
) {
  return te.run(mapper(entity, pack, role, aclAllow, typeMapper), template);
}

export interface MapperOutput {
  name: string;
  plural: string;
  pluralEntry: string;
  singularEntry: string;
  unique: string;
}

// tslint:disable:o-unused-variable
import {
  getFieldsForAcl,
  identityFields,
  getFields,
  idField,
  memoizeEntityMapper,
} from '../../queries';

export const mapper = memoizeEntityMapper(template, _mapper);

export function _mapper(
  entity: Entity,
  pack: ModelPackage,
  role: string,
  aclAllow,
  typeMapper: { [key: string]: (i: FieldType) => string },
): MapperOutput {
  let ids = getFields(entity).filter(idField);
  const mapToGQLTypes = typeMapper.graphql;

  let unique = [
    ...ids.map(f => ({
      name: f.name,
      type: 'ID',
    })),
    ...getFieldsForAcl(role, pack)(aclAllow, entity).filter(identityFields),
  ]
    .map(f => ({
      name: f.name,
      type: mapToGQLTypes(f.type),
    }))
    .map(i => `${i.name}: ${i.type}`)
    .join(', ');

  return {
    name: entity.name,
    plural: entity.plural,
    singularEntry: inflect.camelize(entity.name, false),
    pluralEntry: inflect.camelize(entity.plural, false),
    unique,
  };
}
