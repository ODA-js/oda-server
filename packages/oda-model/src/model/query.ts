import clean from '../lib/json/clean';
import { FieldArgs, QueryInput, QueryStorage } from './interfaces';
import { ModelBase } from './modelbase';
import { IQuery } from './interfaces/model';

export class Query extends ModelBase implements IQuery {
  protected $obj!: QueryStorage;

  public get args(): FieldArgs[] {
    return this.$obj.args;
  }

  public get payload(): FieldArgs[] {
    return this.$obj.payload;
  }

  public updateWith(obj: QueryInput) {
    if (obj) {
      super.updateWith(obj);

      const result = { ...this.$obj };

      let args = obj.args;

      let payload = obj.payload;

      result.args = args;

      result.payload = payload;

      this.$obj = result;
    }
  }

  // it get fixed object
  public toObject() {
    let props = this.$obj;
    let res = super.toObject();
    return clean({
      ...res,
      args: props.args ? props.args : undefined,
      payload: props.payload ? props.payload : undefined,
    });
  }
}
