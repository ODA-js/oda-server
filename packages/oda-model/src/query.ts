import { merge } from 'lodash';
import { Internal } from './element';
import {
  AsHash,
  MetaModelType,
  Nullable,
  assignValue,
  NamedArray,
  MapToArray,
  EntityType,
  ScalarTypeNames,
  SimpleModelTypes,
} from './types';
import { outputPayload } from './utils/converters';
import { inputArgs, inputPayload } from './utils/converters';
import {
  IModelBase,
  ModelBaseInternal,
  ModelBaseInput,
  ModelBase,
  ModelBaseOutput,
  ModelBaseMetaInfo,
} from './modelbase';
import decapitalize from './lib/decapitalize';
import { IObjectTypeField, ObjectTypeFieldInput } from './objecttypefield';
import { IObjectType, ObjectTypeInput } from './objecttype';

export interface IQuery
  extends IModelBase<QueryMetaInfo, QueryInput, QueryOutput> {
  /**
   * set of arguments
   */
  readonly args: Map<string, IObjectType | IObjectTypeField>;
  /**
   * set of output fields
   */
  readonly payload:
    | SimpleModelTypes
    | EntityType
    | IObjectType
    | Map<string, IObjectType | IObjectTypeField>;
}

export interface QueryMetaInfo extends ModelBaseMetaInfo {
  acl: {
    execute: string[];
  };
}

export interface QueryInternal extends ModelBaseInternal {
  args: Map<string, IObjectType | IObjectTypeField>;
  payload:
    | SimpleModelTypes
    | EntityType
    | IObjectType
    | Map<string, IObjectType | IObjectTypeField>;
}

export interface QueryInput extends ModelBaseInput<QueryMetaInfo> {
  args:
    | AsHash<ObjectTypeFieldInput | ObjectTypeInput>
    | NamedArray<ObjectTypeFieldInput | ObjectTypeInput>;
  payload:
    | ScalarTypeNames
    | SimpleModelTypes
    | EntityType
    | ObjectTypeInput
    | AsHash<ObjectTypeFieldInput | ObjectTypeInput>
    | NamedArray<ObjectTypeFieldInput | ObjectTypeInput>;
}

export interface QueryOutput extends ModelBaseOutput<QueryMetaInfo> {
  args: NamedArray<ObjectTypeFieldInput>;
  payload:
    | SimpleModelTypes
    | EntityType
    | ObjectTypeInput
    | NamedArray<ObjectTypeFieldInput | ObjectTypeInput>;
}

export const queryDefaultMetaInfo = {
  acl: {
    execute: [],
  },
};
export const queryDefaultInput = { metadata: queryDefaultMetaInfo };

export class Query
  extends ModelBase<QueryMetaInfo, QueryInput, QueryInternal, QueryOutput>
  implements IQuery {
  public get modelType(): MetaModelType {
    return 'query';
  }

  constructor(init: QueryInput) {
    super(merge({}, queryDefaultInput, init));
  }

  public get args(): Map<string, IObjectType | IObjectTypeField> {
    return this[Internal].args;
  }

  public get payload():
    | SimpleModelTypes
    | EntityType
    | IObjectType
    | Map<string, IObjectType | IObjectTypeField> {
    return this[Internal].payload;
  }

  public updateWith(input: Nullable<QueryInput>) {
    super.updateWith(input);

    assignValue<QueryInternal, QueryInput, NonNullable<QueryInput['name']>>({
      src: this[Internal],
      input,
      field: 'name',
      effect: (src, value) => (src.name = decapitalize(value)),
      required: true,
    });

    assignValue<QueryInternal, QueryInput, NonNullable<QueryInput['args']>>({
      src: this[Internal],
      input,
      field: 'args',
      effect: (src, value) => (src.args = inputArgs(value)),
      required: true,
      setDefault: src => (src.args = new Map()),
    });

    assignValue<QueryInternal, QueryInput, NonNullable<QueryInput['payload']>>({
      src: this[Internal],
      input,
      field: 'payload',
      effect: (src, value) => (src.payload = inputPayload(value)),
      required: true,
      setDefault: src => (src.payload = new Map()),
    });
  }

  public toObject(): QueryOutput {
    const internal = this[Internal];
    const payload = outputPayload(internal.payload);
    return merge({}, super.toObject(), {
      args: MapToArray(this[Internal].args, (_name, value) => value.toObject()),
      payload,
    } as Partial<QueryOutput>);
  }
  public mergeWith(payload: Nullable<QueryInput>) {
    super.mergeWith(payload);
  }
}
