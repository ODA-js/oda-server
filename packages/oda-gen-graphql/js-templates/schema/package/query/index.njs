<#@ chunks "$$$main$$$" -#>
<#@ context 'pkg' #>
<#@ alias 'queries/index' #>

<#- chunkStart(`./queries/index.ts`); -#>
<# pkg.queries.forEach(s=>{#>
import #{s.name} from './#{s.name}';
<#})#>
import { Schema } from '../common';

export default new Schema ({
  name: '#{pkg.name}.queries',
items:[<# pkg.queries.forEach(s=>{#>
 #{s.name},
<#})#>],
});


<# pkg.queries.forEach(s=>{#>
<#- chunkStart(`./queries/${s.name}/index.ts`); -#>
import { Schema } from '../../common';
import gql from 'graphql-tag';
import query from './query';
import input from './input';
import payload from './payload';

export default new Schema({
  name: '#{s.name}.query',
  items:[query, input, payload],
})

<#- chunkStart(`./queries/${s.name}/query.ts`); -#>
import { Query, logger } from '../../common';
import gql from 'graphql-tag';

export default new Query({
  schema:gql`
    extend type RootQuery {
      #{s.name}(input: #{s.name}Input!): #{s.name}Payload
    }
  `,
  resolver:    async (
    owner,
    args: {
  <#- for (let f of s.args) {#>
      #{f.name}?: #{f.type.ts},
  <#-}#>
    },
    context,
    info
  ) => {
    logger.trace('#{s.name}');
    let result: {
      // what must be in output
  <#- for (let f of s.payload) {#>
      #{f.name}?: any; // #{f.type.ts},
  <#-}#>
    };
    result = {};
    // put your code here
    return result;
  }
})

<#- chunkStart(`./queries/${s.name}/input.ts`); -#>

import { Input } from '../../common';
import gql from 'graphql-tag';

export default new Input({
  schema: gql`
    input #{s.name}Input {
    <# for(let arg of s.args){-#>
      #{arg.name}: #{arg.type.gql}
    <#}-#>
    }
  `,
});



<#- chunkStart(`./queries/${s.name}/payload.ts`); -#>
import { Type } from '../../common';
import gql from 'graphql-tag';

export default new Type({
  schema: gql`
    type #{s.name}Payload {
    <# for(let arg of s.payload){-#>
      #{arg.name}: #{arg.type.gql}
    <#}-#>
    }
  `,
});

<#})#>
