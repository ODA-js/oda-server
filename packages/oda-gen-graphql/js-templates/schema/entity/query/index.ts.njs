<#@ chunks "$$$main$$$" -#>
<#@ alias 'query-index'#>
<#@ context 'ctx'#>

<#-chunkStart(`./query/item/index.ts`); -#>
import #{ctx.entry.singularEntry} from './#{ctx.entry.singularEntry}';
import #{ctx.entry.singularEntry}UniqueKeys from './#{ctx.entry.singularEntry}UniqueKeys';
import { Schema } from '../../../../common';

export default new Schema({
  name: '#{ctx.entry.name}.queries.single',
  items: [#{ctx.entry.singularEntry}, #{ctx.entry.singularEntry}UniqueKeys],
});

<#-chunkStart(`./query/list/index.ts`); -#>
#{slot('import-query-list-index-slot')}
import { Schema } from '../../../../common';
export default new Schema({
  name: '#{ctx.entry.name}.queries.list',
  items: [
    #{slot('item-query-list-index-slot')}
  ],
});

<#-chunkStart(`./query/index.ts`); -#>

import list from './list';
import item from './item';
import filters from './filters';

import { Schema } from '../../../common';

export default new Schema({
  name: '#{ctx.entry.name}.query',
  items: [list, item, filters],
});

<#- chunkEnd(); -#>

#{partial(ctx,'query-single')}
#{partial(ctx,'query-list')}
#{partial(ctx.sortOrder,'query-list-sort-order')}
#{partial(ctx.filter,'query-list-complex-filter')}
#{partial(ctx.filter,'query-filters')}
