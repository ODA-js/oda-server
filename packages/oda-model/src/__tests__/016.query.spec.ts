import 'jest';
import { Query, queryDefaultMetaInfo } from '../query';

describe('Query', () => {
  it('default', () => {
    const res = new Query({
      name: 'GetAllUsersThat',
      args: [{ name: 'disabled', type: 'boolean', multiplicity: 'one' }],
      payload: [{ name: 'id', required: true, multiplicity: 'many' }],
    });
    expect(res.metadata).toMatchObject(queryDefaultMetaInfo);
    expect(res.toObject()).toMatchSnapshot('toObject');
  });
});
