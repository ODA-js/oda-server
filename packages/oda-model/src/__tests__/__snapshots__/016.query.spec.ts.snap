// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query default: toObject 1`] = `
Object {
  "args": Array [
    Object {
      "description": undefined,
      "metadata": Object {
        "required": false,
      },
      "multiplicity": "one",
      "name": "disabled",
      "title": undefined,
      "type": Object {
        "multiplicity": "one",
        "name": "Boolean",
        "type": "scalar",
      },
    },
  ],
  "description": undefined,
  "metadata": Object {
    "acl": Object {
      "execute": Array [],
    },
  },
  "name": "getAllUsersThat",
  "payload": Array [
    Object {
      "description": undefined,
      "metadata": Object {
        "required": true,
      },
      "multiplicity": "many",
      "name": "id",
      "title": undefined,
      "type": Object {
        "name": "String",
        "type": "scalar",
      },
    },
  ],
  "title": undefined,
}
`;
